### PR TITLE
Make nuget future proof for download count exceeding int.MaxValue

### LINKED ISSUE #10715: [Speculative bug] Ensure all code paths work with a download count exceeding int.MaxValue
## Details about Problem

NuGet product used: all
Product version: latest

## Repro steps and/or sample project

See https://github.com/NuGet/NuGetGallery/issues/8476. In short, sooner or later Newtonsoft.Json will exceed `int.MaxValue` (2,147,483,647) as may cause problems.

End-to-end, we (NuGet team) need to ensure that a download count exceeding `int.MaxValue` does not break the client. The aforementioned link tracks the server-side investigation/fixes.

I have confirmed the following with Fiddler on client side.

- PM UI Browse tab, V3 nuget.org
  - ✔️ Works up to 1000 trillion (so we have plenty of runway 😄)
  - image
  - ❌ Crashes VS after 1000 trillion
- PM UI Browse tab, V2 nuget.org
  - Same as V3.
- nuget.exe search, V3 nuget.org
  - ✔️ Works up to `long.MaxValue`
  - image
  - ❌ Fails over to USSC (???) after `long.MaxValue`
- nuget.exe search V2 nuget.org
  - ✔️ Works up to `long.MaxValue`
  - image
  - ✔️ Shows N/A after `long.MaxValue` (I think this is a good behavior)
  - image

From this analysis it seems that client is fine for the `int.MaxValue`. However I may be missing places and I think it would be prudent to do a pass over all of the places to make sure at least `int.MaxValue + 1` doesn't crash NuGet since this is a real situation on the horizon.

I don't know all of the places download count is used on the client.

### PR DESCRIPTION
**Pull request by erdembayar**
Fixes: https://github.com/NuGet/Home/issues/10715

Regression? Last working version: N/A

## Description
Mostly we're good, but found 3 places which still using `int` instead of `long`. I only fixed `DownloadCountToVisibilityConverter` and `VersionInfoConverter`. 
I used `downloadCount`/`totalDownloads` as search keyword, I hope I covered all  cases.

1. For count larger than int.MaxValue `DownloadCountToVisibilityConverter.cs` always return collapse.
2. Without this change `VersionInfoConverter` would throw exception like below:

image

3. But last one is public api and V2 protocol so I decided to leave as it's, at least it doesn't throw exception.
https://github.com/NuGet/NuGet.Client/blob/5138176d007034f96afce7132f034403a3888c44/src/NuGet.Core/NuGet.Protocol/LegacyFeed/V2FeedPackageInfo.cs#L186-L188

For sake of completeness I include some results.
image

image

image

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/PackageSearchResourceV3Tests.cs [hunk 1]
```diff
@@ -111,6 +111,7 @@ public async Task PackageSearchResourceV3_GetMetadataAsync_NotFound()
         public async Task PackageSearchResourceV3_GetMetadataAsync_VersionsDownloadCount()
         {
             // Arrange
+            long largerThanIntMax = 2147483657; // int.MaxValue + 10
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/PackageSearchResourceV3Tests.cs]:
```suggestion
            long largerThanIntMax = (long)int.MaxValue + 10;
```

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/PackageSearchResourceV3Tests.cs [hunk 2]
```diff
@@ -137,6 +138,7 @@ public async Task PackageSearchResourceV3_GetMetadataAsync_VersionsDownloadCount
             Assert.Equal(28390569, package.DownloadCount);
             Assert.Equal(14, versions.Count());
             Assert.Equal(64099, versions.First().DownloadCount);
+            Assert.Equal(largerThanIntMax, versions.Skip(1).First().DownloadCount);
```
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/PackageSearchResourceV3Tests.cs]: Change `versions` to include `.ToList()`, then

```suggestion
            Assert.Equal(14, versions.Count);
            Assert.Equal(64099, versions[0].DownloadCount);
            Assert.Equal(largerThanIntMax, versions[1].DownloadCount);
```

**Review by zivkan — APPROVED** [review]

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Converters/DownloadCountToVisibilityConverterTests.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        [Theory]
+        [InlineData(null)]
+        [InlineData(-1)]
+        public void Convert_ValidLongValue_Return_VisibilityCollapsed(long? downloadCount)
+        {
+            var converter = new DownloadCountToVisibilityConverter();
+
+            var converted = converter.Convert(
```
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Converters/DownloadCountToVisibilityConverterTests.cs]: explicit type

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Converters/DownloadCountToVisibilityConverterTests.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        [Theory]
+        [InlineData(0)]
+        [InlineData(1)]
+        [InlineData(2147483657)]
+        public void Convert_ValidLongValue_Return_VisibilityVisible(long? downloadCount)
+        {
+            var converter = new DownloadCountToVisibilityConverter();
+
+            var converted = converter.Convert(
```
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Converters/DownloadCountToVisibilityConverterTests.cs]: explicit type

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/PackageSearchResourceV3Tests.cs (hunk 2 above)
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/PackageSearchResourceV3Tests.cs]: The test seems a little unintuitive. You have to know that `EntityFrameworkSearch` has a 2nd version with download count that was too large for an `int32`. Can you add a comment (I prefer block comments above the method) that somehow paraphrases this is a dependency and one of the assertions in this test?

I suggested a comment rather than an Theory or a separate test, as I expect the perf cost would be higher for this one.

Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/PackageSearchResourceV3Tests.cs]: I just added inline comment here, not above the method, because I have seen before with other cases testing edge cases in middle of other asserts.

Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/PackageSearchResourceV3Tests.cs]: Thank you for review. Taken the suggestion.

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/PackageSearchResourceV3Tests.cs (hunk 1 above)
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/PackageSearchResourceV3Tests.cs]: Sure.

**Review by donnie-msft — APPROVED** [review]
Thanks for being proactive and jumping on this!

Did you happen to test the PM UI with a very large download count to make sure it renders well? We may need separate issue/PR for the XAML.

**Review by nkolev92 — APPROVED** [review]

**Comment by erdembayar** [comment]
> Thanks for being proactive and jumping on this!
> 
> Did you happen to test the PM UI with a very large download count to make sure it renders well? We may need separate issue/PR for the XAML.

@donnie-msft 
I tested some cases already, not sure if need more tests. See below cases, please let me know if we still need  issue/PR for the XAML.

image

image

image

**Comment by donnie-msft** [comment]
> I tested some cases already, not sure if need more tests. See below cases, please let me know if we still need issue/PR for the XAML

That looks good. I don't think we currently render it anywhere else so we should be good.

> But last one is public api and V2 protocol so I decided to leave as it's, at least it doesn't throw exception.

If I understand, this will lead to `0 downloads`? How impactful is that?

**Comment by erdembayar** [comment]
> > But last one is public api and V2 protocol so I decided to leave as it's, at least it doesn't throw exception.
> 
> If I understand, this will lead to `0 downloads`? How impactful is that?

Yes, it'll lead to `0 downloads`. Since it's public api better not to make any changes. Not sure if it's big deal. First of all V2 is kind of outdated protocol. If customer experience problem we better to recommend them use V3.
@joelverhagen  Do you have any concern about V2 is being not fixed?

Update:
Please note either case not good result.
Case 1: Fix V2 protocol api then it's we're breaking public api for some people.
Case 2: Don't fix V2 protocol api then some people starts seeing `0` after few years.
Hopefully everyone using V2 move to V3 around that time.

**Comment by joelverhagen** [comment]
I don't know the code paths that use the V2 model. I have no data to assess impact.

**Comment by erdembayar** [comment]
Synced offline, if anyone using V2 then need to move into V3 since V2 end point is deprecated.
